File: Even Odd Tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isEvenOddTree(root *TreeNode) bool {
    levels := make([][]int, 0)
    var exploreNode func(node *TreeNode, depth int) bool
    exploreNode = func(node *TreeNode, depth int) bool {
        if node == nil {
            return true
        }

        if len(levels) == depth {
            arr := make([]int, 0)
            levels = append(levels, arr)
        }

        if len(levels[depth]) > 0 {
            last_num := levels[depth][len(levels[depth])-1]
            if depth & 1 == 1 {
                if last_num <= node.Val {
                    return false
                }
            } else {
                if last_num >= node.Val {
                    return false
                }
            }
        }

        if depth & 1 == 1 {
            if node.Val & 1 == 1 {
                return false
            }
        } else {
            if node.Val & 1 == 0 {
                return false
            }
        }

        levels[depth] = append(levels[depth], node.Val)
        return exploreNode(node.Left, depth+1) && exploreNode(node.Right, depth+1)
    }
    return exploreNode(root, 0)
}

File: Even Odd Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isEvenOddTree(self, root: Optional[TreeNode]) -> bool:
        levels = []

        def exploreNode(node, depth):
            nonlocal levels
            if node is None:
                return True

            if len(levels) == depth:
                levels.append([])

            if len(levels[depth]) > 0:
                last_num = levels[depth][-1]
                if depth & 1:  # Odd level
                    if last_num <= node.val:  # Not decreasing
                        return False
                else:
                    if last_num >= node.val:  # Not increasing
                        return False

            if depth & 1:  # Odd level
                if node.val & 1:  # Odd number
                    return False
            else:  # Even level
                if node.val & 1 == 0:  # Even number
                    return False

            levels[depth].append(node.val)
            return exploreNode(node.left, depth + 1) and exploreNode(node.right, depth + 1)

        return exploreNode(root, 0)
